Source/WebKit/Shared/cf/CoreIPCSecKeychainItem.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if HAVE(SEC_KEYCHAIN)
+
+#import <Security/SecKeychainItem.h>
+#import <wtf/ProcessPrivilege.h>
+#import <wtf/RetainPtr.h>
+
+namespace WebKit {
+
+class CoreIPCSecKeychainItem {
+public:
+    CoreIPCSecKeychainItem(SecKeychainItemRef keychainItem)
+        : m_persistentRef(persistentRefForKeychainItem(keychainItem))
+    {
+    }
+
+    CoreIPCSecKeychainItem(RetainPtr<CFDataRef> data)
+        : m_persistentRef(data)
+    {
+    }
+
+    CoreIPCSecKeychainItem(const IPC::DataReference& data)
+        : m_persistentRef(adoptCF(CFDataCreate(kCFAllocatorDefault, data.data(), data.size())))
+    {
+    }
+
+    RetainPtr<SecKeychainItemRef> createSecKeychainItem() const
+    {
+        RELEASE_ASSERT(hasProcessPrivilege(ProcessPrivilege::CanAccessCredentials));
+
+        SecKeychainItemRef keychainItem = NULL;
+        CFDataRef data = m_persistentRef.get();
+        ASSERT(data);
+
+        // SecKeychainItemCopyFromPersistentReference() cannot handle 0-length CFDataRefs.
+        if (CFDataGetLength(data)) {
+            ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+            SecKeychainItemCopyFromPersistentReference(data, &keychainItem);
```

Comment:
There aren't any API/SPI to manually deconstruct a `SecKeychainItem` into component parts. Fortunately, a persistent reference is like an alias or PK to the original item—not the item itself—so there _should_ be less room for mischief with attacker data.